File: lib/checksum.js
/**
 * Deterministic file checksum for Aigis content artifacts.
 *
 * Algorithm (locked, documented so it's reproducible from this spec):
 *   1. Read the file as UTF-8 bytes.
 *   2. Strip a leading UTF-8 BOM (\uFEFF) if present.
 *   3. Normalize line endings: convert CRLF and lone CR to LF.
 *   4. SHA-256 the resulting bytes.
 *   5. Return the lowercase hex digest.
 *
 * Why canonicalize: the checksum gets written into IDE rules files and
 * compared on subsequent invocations of `aigis init`. Editors / git-attribute
 * settings can flip line endings or insert BOMs without the user noticing,
 * which would otherwise produce spurious mismatches.
 */

const fs = require('fs');
const crypto = require('crypto');

function canonicalizeBytes(text) {
  // Strip leading BOM
  if (text.charCodeAt(0) === 0xFEFF) {
    text = text.slice(1);
  }
  // CRLF and lone CR → LF
  return text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
}

function computeChecksum(text) {
  const canonical = canonicalizeBytes(text);
  return crypto.createHash('sha256').update(canonical, 'utf8').digest('hex');
}

function computeFileChecksum(filepath) {
  const text = fs.readFileSync(filepath, 'utf8');
  return computeChecksum(text);
}

module.exports = { computeChecksum, computeFileChecksum, canonicalizeBytes };

File: cli-python/aigis_cli/checksum.py
"""Mirror of lib/checksum.js. Spec is the same: strip BOM, normalize CRLF/CR to LF, SHA-256."""

from __future__ import annotations

import hashlib
from pathlib import Path


def canonicalize(text: str) -> str:
    if text.startswith("\ufeff"):
        text = text[1:]
    return text.replace("\r\n", "\n").replace("\r", "\n")


def compute_checksum(text: str) -> str:
    canonical = canonicalize(text)
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def compute_file_checksum(path: str | Path) -> str:
    text = Path(path).read_text(encoding="utf-8")
    return compute_checksum(text)
